### PR TITLE
Support Tizen GBS build

### PR DESCRIPTION
**Pull request by sumin-lim**
GBS build script, its running script and guide are added.
IoT.js-DCO-1.0-Signed-off-by: Sumin Lim sumin.lim@samsung.com

### CODE REVIEW

File: config/tizen/gbsbuild.sh [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+
+# Copyright 2016-present Samsung Electronics Co., Ltd. and other contributors
```
Comment (yichoi) [inline config/tizen/gbsbuild.sh]: 2017-present ?

File: config/tizen/gbsbuild.sh [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+read -p "The folder, '.git' will be removed. \
+Are you sure to continue? (y/n) " -n 1 -r
+echo    # (optional) move to a new line
+if [[ $REPLY =~ ^[Yy]$ ]]
+then
+
+echo `pwd`
```
Comment (LaszloLango) [inline config/tizen/gbsbuild.sh]: Wrong indentation

**Comment by zherczeg** [comment]
Makes sense although I cannot try it.

File: config/tizen/gbsbuild.sh [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+echo "set(CMAKE_SYSTEM_NAME Linux)
+set(CMAKE_SYSTEM_PROCESSOR armv7l-el)
+set(FLAGS_COMMON_ARCH -mlittle-endian -mthumb)" \
+>> ./deps/jerry/cmake/toolchain_linux_armv7l-el.cmake
+
+echo "3 files have been modified."
+
+rm -rf .git/
+# Initialize Git repositoryㅣ
+if [ ! -d .git ]; then
```
Comment (LaszloLango) [inline config/tizen/gbsbuild.sh]: Please use a consistent style in the file.
```bash
if [ ]
then
  # 2 or 4 space indentation
fi
```

or

```bash
if [ ]; then
  # 2 or 4 space indentation
fi
```

File: config/tizen/gbsbuild.sh (hunk 2 above)
Comment (sumin-lim) [inline config/tizen/gbsbuild.sh]: Done.

File: config/tizen/gbsbuild.sh (hunk 3 above)
Comment (sumin-lim) [inline config/tizen/gbsbuild.sh]: Done, I used first one.

File: config/tizen/gbsbuild.sh [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    git init ./
+    git checkout -b tizen_gbs
+    git add ./
+    git commit -m "Initial commit"
+  fi
+
+  echo "Calling core gbs build command"
+  gbscommand="gbs build -A armv7l --include-all --clean"
+  echo $gbscommand
+  if eval $gbscommand; then
```
Comment (LaszloLango) [inline config/tizen/gbsbuild.sh]: Inconsistent style and indentation (below). Please fix this too.

File: config/tizen/packaging/iotjs.spec [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+%description devel
+Development libraries for %{name}
+
+%prep
+%setup -q -c
+chmod g-w %_sourcedir/*
+cat LICENSE
+cp %{SOURCE1001} .
+touch abc.txt
```
Comment (LaszloLango) [inline config/tizen/packaging/iotjs.spec]: Why is this needed?

File: config/tizen/packaging/iotjs.spec [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+%build
+./tools/build.py --clean --buildtype=%{build_mode} --target-arch=arm \
+ --target-os=tizen --target-board=artik10 \
+ --no-init-submodule --no-parallel-build --no-check-test
```
Comment (LaszloLango) [inline config/tizen/packaging/iotjs.spec]: I don't see where are dependencies resolved. Are you sure this going to work with `--no-init-submodule`?

File: config/tizen/gbsbuild.sh (hunk 4 above)
Comment (sumin-lim) [inline config/tizen/gbsbuild.sh]: Done.

File: config/tizen/packaging/iotjs.spec (hunk 6 above)
Comment (sumin-lim) [inline config/tizen/packaging/iotjs.spec]: Yes, I'm sure. GBS can not make update with git submodules on build.py is running. That's why .git folder removed. You can try its build with this file(copy this on your $HOME, ~/) and 
make a script copy (./config/tizen/.gbsbuild.sh .) on the working.
You can see this script pull the submodule before removing .git folder. 

https://gist.githubusercontent.com/sumin-lim/fcbfdae32f76954e462dd615f8d93e91/raw/677ae88a23f51b9c89ba0ec234843f3e89b41508/.gbs.conf

**Comment by pmarcinkiew** [comment]
I tested this commit with GBS at Ubuntu and I managed to compile valid RPM packaged. There is no way to integrated directly GBS compilation with Travis. No-init-submodule is mandatory for GBS because this environment doesn't provide access to working GIT.

**Comment by yichoi** [comment]
This is for the Tizen IoT Headless on RPi3. It means it is for specific target like RPi2/Linux, STM32F4/Nuttx so that we don't need to handle this with travis but with [daily performance monitoring](https://samsung.github.io/iotjs-test-results/)

File: config/tizen/packaging/iotjs.spec (hunk 6 above)
Comment (sumin-lim) [inline config/tizen/packaging/iotjs.spec]: Tizen profiles runs on Travis still, not related with GBS. (I changed the code on the review)
It's possible to make Tizen profile travis to create a .rpm. However, It's possible to guarantee as the Tizen OBS system and a maintainer apply iot.js source code on the Tizen.org. So I didn't make it.

For the dependency, let me explain more.
1. The files in './conf/tizen/packaing' are for the GBS building, which runs on the developer's PC and Tizen OBS system to create RPM Package.
2. A file, 'gbsbuild.sh' is only for developer who wants make own IoT.js module or changes on Tizen device.

**Comment by sumin-lim** [comment]
I think removing a folder,'.git' is not a good idea on current working directory.
Therefore, I modified gbsbuild.sh to create a new working directory on upper's new folder, 'iotjs_tizen_gbs'
Please review to this new changes.

File: config/tizen/packaging/iotjs.spec (hunk 5 above)
Comment (sumin-lim) [inline config/tizen/packaging/iotjs.spec]: deleted test code.

File: config/tizen/packaging/iotjs.spec (hunk 6 above)
Comment (LaszloLango) [inline config/tizen/packaging/iotjs.spec]: I see. Thanks for the explanation.

File: config/tizen/gbsbuild.sh [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    git remote add origin https://github.com/samsung/iotjs
+    echo "(You cant fetch origin repository with this command)"
+    echo "   git fetch --all"
+    echo "========================================================"
+    # git remote add origin
+    # https://review.tizen.org/git/platform/upstream/iotjs
+    git branch -a
+    git status
+  else
+      echo "GBS Build failed!"
```
Comment (LaszloLango) [inline config/tizen/gbsbuild.sh]: wrong indentation

File: config/tizen/gbsbuild.sh [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+echo "*                                                                *"
+echo "* Please input user, passwd of http://Tizen.org on '~/.gbs.conf' *"
+echo "* For more information, please read Guide Docs. folder           *"
+echo "* ~/.gbs.conf sample is at 'config/tizen/sample.gbs.conf'.       *"
+echo "*                                                                *"
+echo "******************************************************************"
+read -p "[Warning] This working folder will be copied to ../iotjs_tizen_gbs \
+Are you sure to continue? (y/n) " -n 1 -r
+cd ..
+cp -ra iotjs iotjs_tizen_gbs
```
Comment (daeyeon) [inline config/tizen/gbsbuild.sh]: We can't assume that the previous working directory is always iotjs. So I suggest using `cp -ra $OLDPWD iotjs_tizen_gbs`

File: config/tizen/gbsbuild.sh [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+  set(CMAKE_SYSTEM_NAME Linux)
+  set(CMAKE_SYSTEM_PROCESSOR arm)"\
+   >> ./deps/libtuv/cmake/config/config_arm-tizen.cmake
+
+  rm ./deps/jerry/cmake/toolchain_linux_armv7l-el.cmake
+  echo "set(CMAKE_SYSTEM_NAME Linux)
+  set(CMAKE_SYSTEM_PROCESSOR armv7l-el)
+  set(FLAGS_COMMON_ARCH -mlittle-endian -mthumb)" \
+  >> ./deps/jerry/cmake/toolchain_linux_armv7l-el.cmake
```
Comment (daeyeon) [inline config/tizen/gbsbuild.sh]: Please make the indent size before `>>` the same as above.

File: config/tizen/gbsbuild.sh (hunk 8 above)
Comment (sumin-lim) [inline config/tizen/gbsbuild.sh]: Done, also I moved these lines to after 'then'

File: config/tizen/gbsbuild.sh (hunk 9 above)
Comment (sumin-lim) [inline config/tizen/gbsbuild.sh]: indent size is 2+1. After '\' one more space on next line.

File: config/tizen/gbsbuild.sh (hunk 7 above)
Comment (sumin-lim) [inline config/tizen/gbsbuild.sh]: Done.

**Comment by yichoi** [comment]
check the build log please
https://travis-ci.org/Samsung/iotjs/jobs/252693791

**Comment by sumin-lim** [comment]
TC "test_net_2.js",(using socket) often fails on a assert statement.  It's not related with this PR.
